### PR TITLE
TRT 4.0 update

### PR DESCRIPTION
**Pull request by jjsjann123**
updated feature support for TRT 4.0 layers
  disabled broken shape inference (added TODO)
  removed unused code
  code is compatible with TRT 3.0.4

### CODE REVIEW

**Comment by jjsjann123** [comment]
Ping @aaroey @benbarsdell @samikama

**Review by aaroey — CHANGES_REQUESTED** [review]
High level comment: please add unit tests to verify that each converter is doing the right thing. I know this is hard, especially when there are no tests done before. But it would be even harder later if we don't do it today, and the lack of test will make future development break things easily. So, please follow the convention and add tests.

These are just initial comments, more will come.
Thanks.

**Comment by aaroey** [comment]
@jjsjann123 Please add the tests in this PR then, I'm happy to review them at the same time.

**Comment by jjsjann123** [comment]
even better! Will push it later today.
just FYI, unit tests are totally separate from the code so there's no need to hold back the review waiting for my push

really appreciate the review :)

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 1]
```diff
@@ -53,6 +53,31 @@ limitations under the License.
 //  would work!
 #define CHECK_EQ_TYPE(val1, val2) CHECK_EQ((int)val1, (int)val2)
 
+#define TFTRT_RETURN_ERROR_IF_FALSE(ptr, node)                               \
+  do {                                                                       \
+    if (ptr == false) {                                                      \
+      return tensorflow::errors::Internal(string("TFTRT::") + __FUNCTION__ + \
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Manually separated the lines.
But seems to be violating clang-format. (The automatic correction put the lines back)

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 2]
```diff
@@ -75,13 +102,108 @@ inline tensorflow::Status ConvertDType(tensorflow::DataType tf_dtype,
...
+  std::memcpy(r_s + max_d - operand_r.nbDims, operand_r.d,
+              operand_r.nbDims * element_size);
+
+  // set -1 for batch dimension, since batch size is not supposed to be
+  // broadcasted
+  if (operand_l_is_tensor) {
+    if (max_d != l_d) {  // if broadcast beyond batch dimension, fail
+      return false;
+    }
+    l_s[0] = -1;
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: It seems this does not align with what the comment says:
```
+    T: 3 5 1
+    -> T: 1 1 1 -1 3 5 1
```
but here it assigns -1 to first element.

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 3]
```diff
@@ -790,118 +918,23 @@ tensorflow::Status BinaryCompute(const TRT_ShapedWeights& iweights_l,
...
-  return ret;
-}
-
 // TODO(jie): broadcast is needed yet not implemented.
 // Only implemented channel wise for the time being
 tensorflow::Status BinaryTensorOpWeight(
     Converter& ctx, const tensorflow::NodeDef& node_def,
     const nvinfer1::ITensor* tensor, TRT_ShapedWeights weights,
-    std::vector<TRT_TensorOrWeights>* outputs) {
+    std::vector<TRT_TensorOrWeights>* outputs, bool swapped_inputs) {
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Please follow the convention by making all input parameters appear before any output parameter.
Also, what does swapped_inputs means? Please comment.

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 4]
```diff
@@ -910,6 +943,12 @@ tensorflow::Status BinaryTensorOpWeight(
   auto dims_w = weights.shape_;
   auto dims_t = tensor->getDimensions();
 
+  // TODO(jie): addScale checks for input tensor dimension
+  if (dims_t.nbDims != 3) {
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Can we put all checks like this into a method and call it in the segmenter, so we can avoid putting nodes into the trt engine at the very beginning, instead of delaying the check until here?

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 5]
```diff
@@ -1056,6 +1123,17 @@ tensorflow::Status ConvertConv2DHelper(
   VLOG(2) << "groups count: " << num_groups;
 
   TRT_ShapedWeights weights_rsck = inputs.at(1).weights();
+
+  VLOG(2) << "weight shape: " << weights_rsck.shape_.nbDims;
+  for (int i = 0; i < weights_rsck.shape_.nbDims; i++) {
+    VLOG(2) << weights_rsck.shape_.d[i];
+  }
+
+  if (weights_rsck.shape_.nbDims != 4) {
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Same question about putting the checks in segmenter.

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 6]
```diff
@@ -1147,18 +1232,71 @@ tensorflow::Status ConvertConv2DHelper(
                                            node_def.name());
 }
 
+bool PrepareTensorForShape(Converter& ctx, const TRT_TensorOrWeights& input,
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Please fix the order of input and output parameters here and all other places in this PR.

Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Please comment what this method does.

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 7]
```diff
@@ -1147,18 +1232,71 @@ tensorflow::Status ConvertConv2DHelper(
...
+    return false;
+#endif
+  }
+  return true;
+}
+
 tensorflow::Status BinaryTensorOpTensor(
     Converter& ctx, const tensorflow::NodeDef& node_def,
-    const nvinfer1::ITensor* tensor_l, const nvinfer1::ITensor* tensor_r,
+    const TRT_TensorOrWeights operand_l, const TRT_TensorOrWeights operand_r,
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: const xx&?

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 8]
```diff
@@ -1468,39 +1608,39 @@ tensorflow::Status ConvertConst(Converter& ctx,
...
-            .end());  //  make a local copy first to flatten
-    memcpy(dst, tensor_data.data(), len_data);  // store into weight store
+    if (weights_tensor.float_val_size() == 1) {
+      std::fill_n((float*)dst, GetShapeSize(scalar_shape),
+                  *weights_tensor.float_val().begin());
+    } else {
+      std::vector<float> tensor_data(
+          weights_tensor.float_val().begin(),
+          weights_tensor.float_val()
+              .end());  //  make a local copy first to flatten
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Please put the comment on a separate line above the code line.

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 9]
```diff
@@ -1520,11 +1660,17 @@ tensorflow::Status ConvertConst(Converter& ctx,
...
-                                          //  doesn't have to be contigous
-    memcpy(dst, tensor_data.data(), len_tensor);  // store into weight store
+    if (weights_tensor.int_val_size() == 1) {
+      std::fill_n((int*)dst, GetShapeSize(scalar_shape),
+                  *weights_tensor.int_val().begin());
+    } else {
+      std::vector<int32> tensor_data(
+          weights_tensor.int_val().begin(),
+          weights_tensor.int_val()
+              .end());  //  make a local copy first to flatten
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Same here.

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 10]
```diff
@@ -1694,10 +1895,82 @@ tensorflow::Status ConvertReduce(Converter& ctx,
     // Apply permutation before extracting dimension for pool_kernel
     output_tensor = ctx.TransposeTensor(
         const_cast<nvinfer1::ITensor*>(output_tensor), permutation_order);
+    TFTRT_RETURN_ERROR_IF_NULLPTR(output_tensor, node_def.name());
   }
   outputs->push_back(TRT_TensorOrWeights(output_tensor));
   return tensorflow::Status::OK();
 }
+#endif
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: `#elseif NV_TENSORRT_MAJOR > 3`

File: tensorflow/contrib/tensorrt/test/unit_tests/BatchMatMulTest.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+from tensorflow.python.ops import nn_impl
+from tensorflow.python.ops import variable_scope
+from tensorflow.python.ops import variables
+from tensorflow.python.ops import init_ops
+from tensorflow.python.ops import gen_array_ops
+from tensorflow.python.ops import gen_math_ops
+from tensorflow.python.ops import math_ops
+from tensorflow.python.layers import core
+from tensorflow.python.training import training
+from base_unit_test import BaseUnitTest
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/test/unit_tests/BatchMatMulTest.py]: Thanks so much for adding the tests!

Please use absolute path and add corresponding build targets for the tests.
Same for other tests.

File: tensorflow/contrib/tensorrt/test/unit_tests/BatchMatMulTest.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+          dtype=dtypes.float32, shape=self.inp_dims, name="input")
+
+      b = constant_op.constant(
+          np.random.randn(12, 5, 12, 7), dtype=dtypes.float32)
+      x1 = math_ops.matmul(x, b)
+      b = constant_op.constant(
+          np.random.randn(5, 1, 1), dtype=dtypes.float32)
+      x1 = x1 + b
+
+      var = variable_scope.get_variable("test", [12, 5, 12, 7], dtype=dtypes.float32, initializer=init_ops.truncated_normal_initializer)
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/test/unit_tests/BatchMatMulTest.py]: Please fix the formatting for all the tests.

**Review by aaroey — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 13]
```diff
@@ -75,13 +102,108 @@ inline tensorflow::Status ConvertDType(tensorflow::DataType tf_dtype,
...
+  int l_s[max_nb_dims];
+  std::fill(l_s, l_s + max_nb_dims, 1);
+  int l_d = operand_l_is_tensor ? operand_l.nbDims + 1 : operand_l.nbDims;
+  int r_s[max_nb_dims];
+  std::fill(r_s, r_s + max_nb_dims, 1);
+  int r_d = operand_r_is_tensor ? operand_r.nbDims + 1 : operand_r.nbDims;
+
+  int max_d = std::max(l_d, r_d);
+  std::memcpy(l_s + max_d - operand_l.nbDims, operand_l.d,
+              operand_l.nbDims * element_size);
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Yes this is true. Do we want to replace element_size with sizeof(int)?

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc (hunk 2 above)
Comment (jjsjann123) [reply tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: The comment is trying to illustrate why we cannot support this case.
In the code, if we need to promote the rank of the tensor operand, the state if (max_d != l_d) returns true and the function returns false.

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc (hunk 4 above)
Comment (jjsjann123) [reply tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: We have undergoing refactoring work on the segmenter. It is to improve the robustness of the converter and expand checking only the op types to consider attributes as well as input types.

So that will come in as future PRs.

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc (hunk 6 above)
Comment (jjsjann123) [reply tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Did a git diff and fixed two other functions with the same problem :)

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc (hunk 10 above)
Comment (jjsjann123) [reply tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Good catch

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 14]
```diff
@@ -49,9 +49,36 @@ limitations under the License.
...
-//  Check if the types are equal. Cast to int first so that failure log message
-//  would work!
-#define CHECK_EQ_TYPE(val1, val2) CHECK_EQ((int)val1, (int)val2)
+// Check if the types are equal. Cast to int first so that failure log message
+// would work!
+#define TFTRT_CHECK_EQ_TYPE(val1, val2) CHECK_EQ((int)val1, (int)val2)
+
+#define TFTRT_RETURN_ERROR_IF_FALSE(ptr, node) \
+  do {                                         \
+    if (ptr == false) {                        \
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: s/false/nullptr? Is ptr a pointer? If so what's the difference between this macro and the xx_IF_NULLPTR below? Otherwise please rename 'ptr' to something else.

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc [hunk 15]
```diff
@@ -1567,47 +1712,105 @@ tensorflow::Status ConvertBinary(Converter& ctx,
...
-                                inputs.at(0).weights(), outputs);
+  // Try to convert into Scale layer first (for better performance)
+  // Since scale layer supports restricted broadcast policy and op types, we
+  // allow failure and try to handle it through Elementwise op
+  // (BinaryTensorOpTensor)
+  if (inputs.at(0).is_tensor() && inputs.at(1).is_weights()) {
+    auto status = BinaryTensorOpWeight(ctx, node_def, inputs.at(0).tensor(),
+                                       inputs.at(1).weights(), false, outputs);
+#if NV_TENSORRT_MAJOR == 3
+    TF_RETURN_IF_ERROR(status);
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: For performance consideration, we prefer to use BinaryTensorOpWeight ( Scale layer in TRT ). But there are cases that this conversion might fail (transpose, broadcasting on batch, etc), which BinaryTensorOpTensor could catch.

But this is only applicable in TRT 4.0 thanks to the Constant layer added in TRT 4.0.

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc (hunk 14 above)
Comment (jjsjann123) [reply tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: I'll rename ptr.

File: tensorflow/contrib/tensorrt/convert/convert_nodes.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (aaroey) [reply tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Would you please add that as comment? Also, how about:
```
Status status;
if (input 0 is tensor and input 1 is weight) {
  status = BinaryTensorOpWeight(..., false, ...);
} else if (input 0 is weight and input 1 is tensor) {
  status = BinaryTensorOpWeight(..., true, ...);

#if NV_TENSORRT_MAJOR == 3
} else {
#else
}
if ((input 0 is tensor and input 1 is tensor) || !status.ok()) {
#endif

  status = BinaryTensorOpTensor(...);
}
#endif

return status;
```

[the remark this answers is not in the document]
Comment (jjsjann123) [reply tensorflow/contrib/tensorrt/convert/convert_nodes.cc]: Looks good. Will do.

**Comment by aaroey** [comment]
Thanks for the fix. The c++ changes looks good to me. I read the python tests and they have duplicate logic comparing to tf_trt_integration_test.py. I would prefer to remove them in this PR and put them in a new PR, and I can merge them with tf_trt_integration_test.py. What do you think?

File: tensorflow/contrib/tensorrt/test/tf_trt_integration_test.py [hunk 16]
```diff
@@ -124,15 +124,15 @@ def GetMultiEngineGraphDef(dtype=dtypes.float32):
 
 TEST_GRAPHS = {
     "SingleEngineGraph":
-        TestGraph(
-            gdef=GetSingleEngineGraphDef(),
-            num_expected_engines=1,
-            expected_output_dims=(100, 6, 6, 6)),
+    TestGraph(
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/test/tf_trt_integration_test.py]: Please revert the changes to this file.

**Comment by aaroey** [comment]
Thanks so much for the patient, LGTM.

**Review by aaroey — APPROVED** [review]
